### ISSUE TITLE
Key.fingerprint() should replace inline implementations of key fingerprinting

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3499]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3499|
|Type|enhancement|
|Created|2008-10-24 15:49:26Z|

In `SSHClientTransport.ssh_KEX_DH_GEX_GROUP`, `SSHClientTransport.ssh_KEX_DH_GEX_REPLY`, `ckeygen._saveKey`, and `ckeygen.printFingerprint`, there are four different implementations of the same key fingerprinting algorithm.  There's also a re-implementation of the same kind of bytewise hexlification in `Key.__repr__`, although not for a fingerprint.

These should be factored into a single function, ideally `Key.fingerprint()`; perhaps a `keys.hexlify` would be helpful for `Key.__repr__` too.

<details><summary>Searchable metadata</summary>

```
trac-id__3499 3499
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__easy easy
time__1224863366000000 1224863366000000
changetime__1226804104000000 1226804104000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set status to `closed`|
|-|-|

(In [25365]) Fixes for the conch client's known_hosts parsing and host key verification.

Fixes [#10942](https://github.com/twisted/twisted/issues/10942)
Fixes [#10890](https://github.com/twisted/twisted/issues/10890)
Fixes [#3494](https://github.com/twisted/twisted/issues/3494)
Fixes [#3496](https://github.com/twisted/twisted/issues/3496)
Fixes [#10884](https://github.com/twisted/twisted/issues/10884)
Fixes [#3499](https://github.com/twisted/twisted/issues/3499)

The main thrust of this change is that it adds a new module to conch, `twisted.conch.client.knownhosts`, which provides a structured representation of OpenSSH's `known_hosts` file.  This is a big step in the direction of modularizing conch's support for verifying host keys and storing the results of that verification, although the internal connection APIs are mostly unchanged at this point, they could now easily be adjusted to speak to an abstract interface and still manipulate the user's actual known_hosts entries rather than completely overriding them.

The individual bugs which this change fixes are too numerous to bother describing twice; have a look at the tickets themselves more information.  Suffice it to say that the conch client is now a lot more reliable (although it still has a long way to go).

In addition to making the conch client more stable, the new API allows for some programmatic manipulation of known_hosts files, which might be independently useful.

Additionally, although the review notes some ways in which the tests can still be improved, the tests for this module can be cited as a _much_ better example for conch contributors as to how unit tests within conch should look.  While these tests are covering almost exclusively new code, they are covering functionality which has existed (and been broken) for quite some time.

Author: glyph

Reviewers: exarkun, jml

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
